### ISSUE TITLE
SSR broken when served from a server build

### ISSUE THREAD

**Issue by mattkrick** [rofrischmann/react-look#276]
**Version**: 1.0.0 (beta8)
**Environment**: React
**Type**: Bug

-------------------
#### Description
Probably due to the singleton nature of the StyleContainer, SSR doesn't work if the routes come from a webpack build. Related to #267

#### Steps to reproduce
1. git clone https://github.com/mattkrick/react-look
2. cd react-look
3. npm run build:server
4. npm run demo:universal
5. Observe the FOUC

#### Expected Behavior
no FOUC

#### Actual Behavior
a big FOUC

CC @jordanh

**Comment by jordanh** [comment]
Hi @rofrischmann!

Last night @mattkrick & I were investigating #267 on our [Action](https://github.com/ParabolInc/action) codebase. In our universal app, we `require()` the client build into the server. This causes there to be two `StyleContainer` singletons, making the current `react-look` SSR interface challenging to use. (Matt's demonstrates this rather clearly by modifying the `react-look` example app in [this changeset](https://github.com/mattkrick/react-look/commit/052a047048d491798bb1b497e1e733f887bec896))

@mattkrick and I believe in "don't hate, iterate." We ❤️  `react-look` and were thinking about how we might modify the API. I'd like to run something by you and gather your reaction. If you like it, we'd be happy to try an implementation:

   1. Add a second context variable to `<LookRoot>` alongside `_lookConfig` called `_lookRoot`
   2. Add the parameter `styles` to the `look()` component decorator that accepts objects returned by `StyleSheet.create({...})`.
   3. Modify the `look()` decorator so these styles then are added to `<LookRoot>`'s specific `StyleContainer` instance by pulling it from the react context

Caveats:

  * This _won't_ fix `StyleSheet.addCss()` or `StyleSheet.addFont()` in spots where a second `StyleContainer` is at play, but this could be worked around by locating these calls elsewhere in the application

   * This doesn't move you closer to stripping your dependency on React 😉

**Comment by rofrischmann** [comment]
@mattkrick @jordanh I really love how you're tackling the issue. That's what I call true open source. 
First of all I want to point out that I am **not** trying to get rid of React. react-look is a React specific library and will always be. Furthermore I am right now creating some kind of low-level API to handle styles with JavaScript as well as rendering to DOM, ...
I will want to bring this to Look as well to solve some issues as well as API design problems. I had some similar ideas to what you have proposed above. I would love to see those changes in production! This will help at least until my new renderer is up and ready if not at all as I might use it for that as well.

**Comment by jordanh** [comment]
Ah, @rofrischmann, I had misunderstood your comments on #267. Makes sense!

I'll fork the `react-look` codebase and see what we can cook up...

**Comment by mattkrick** [comment]
@rofrischmann, have you opened up lines of communication with the khan academy folks over in https://github.com/Khan/aphrodite? I definitely think there's room for both: Aphrodite fills the barebones, front-end agnostic niche, and react-look fills the react/react-native full-feature niche. 

That established, I think we could share the best ideas between the packages. Checking out how they do SSR (https://github.com/Khan/aphrodite/blob/master/src/index.js#L25), it's pretty eloquent & would solve #275. There may be room to optimize it for react builds, but as is, I think it serves as a good base for a refactor.

**Comment by rofrischmann** [comment]
I have not yet checked aphrodite in detail, but found they're doing great stuff. Sadly it was not available when I started with Look at all. Still I am working on sth. different which might not go well with aphrodite (at least not in every aspect). Though I will study its source code and compare both ideas. Thanks for mentioning!

**Comment by mattkrick** [comment]
what's the different thing your thinking about doing? Just wanna make sure if I make a PR I stay within those lines

**Comment by mattkrick** [comment]
Quick update about possible solutions:

Solution 1: Add styles to the HOC (`@look(styles)`)
This is pretty easy. By passing the styles to a HOC that has access to context, we can build on the variable in the context (eg `_lookConfig`) instead of a singleton that comes from an import. For a working implementation, see: https://github.com/kriasoft/isomorphic-style-loader.

Solution 2: 
Have `addCSS` return a HOC that returns a proxy proxy. For example:

```
StyleContainer.subHOC = StyleSheet.addCSS({...})
StyleContainer.subHOC === function(configInContext) { return new Proxy(StyleContainer.tmpStyles,...)}
```

Then, in the `look` HOC, it connects the subHOC to the context object. Finally, when the component accesses a style, eg `styles.foo`, it determines if `target === styles`, then registers `target[prop]` with the `configInContext`

Pro, no API change. Con, Proxies will require a babel tool to work with browsers <IE11. Not sure what your target is.

Solution 3: Make the user write a function inside their component
This is the route aphrodite took. `<FooComponent styles={css(styles.foo)}/>`.
When that function is called, it puts that single style prop into a buffer & then sends that buffer to the style tag in the DOM. 
This has the added benefit of eliminating styles that are declared but never used. That's nice!
But, again it requires a breaking API change & makes for a slightly clunkier API than solution 1.

Solution 4:
Don't use real CSS. This is the option that radium takes & that's why they can write `@Radium` and don't have this problem. I don't like this option.

**Comment by rofrischmann** [comment]
First of all, thanks for rolling up all those information!
I'd like to give some comments on each method:

**Solution 4**
For me, that's a no-go. Look *was* pure inline before, which still didn't play well as performance was suboptimal and you had to either use fallbacks (with CSS again) for media query, pseudo, ... or implement each on your own (which was a true mess and a lot of code, still not as performant as the CSS equivalent). Also note that with React 15, Radium will face huge problems as assigning strinsgs with fallback values is no longer possible e.g. `display: -webkit-flex;display:flex`.

**Solution 3**
With this solution we would loose a lot of dynamic feature as it only transforms the styles once into classNames and then stays static (if I understood it wrong). How would you add dynamic classNames?

**Solution 2**
This seems to be kind of overload. I don't know, I don't like that at all.

**Solution 1**
Solution 1 is the one I actually like the most. It has a slightly different API, but building on top of React's `context` seems to be the right way. With this solution each `<LookRoot>` would carry its own StyleContainer which also opens some level of optimizations as well.

**Comment by mattkrick** [comment]
@rofrischmann good catch on **solution 3**. To solve it, what we'd have to do is create that `css` function inside the look HOC and attach props & state to its closure. Like you, I still think the first is more eloquent. Plus, that means we could offload a lot of the magic behind `addCSS` and stick it in the HOC, at least for the non-native version (haven't dug into the react-native code yet).

**Comment by mattkrick** [comment]
Extra thoughts: Part of the API that i really like is that `styles` is defined outside of the component, it's not passed in via props. To keep that & eliminate the singleton would be a little bit of a challenge:
- The `StyleSheet.create` method would likely to write to a temporary `StyleContainer`, since we can't trust that the singleton is used everywhere. 
- For that same reason, we can't use an autoincrementing hash, since there might be naming conflicts with other singletons. We could consider something like a murmurhash to hash on the contents of the object (but it's still really fast).
- The `@look` would need to access the temporary `StyleContainer` and then take those values & dump it in a `context._lookConfig.styleContainer`, which would be the style container that has all the values from all the singletons.

After realizing all the work involved, solution 3 looks a little nicer, especially since a similar API already exists for using multiple classes: `c(class1, class2)`. We could instead offer up a function that has the props & state encapsulated, eg `this.props.css`. That function would hash the style, put the values in the `context._lookConfig.styleContainer`, and return the hashed class name. This has the added benefit of only appending styles that are used. The function could also take multiple args, so it would eliminate the need for the `StyleSheet.combine` method.
